nemo/collections/nlp/modules/common/megatron/attention.py

```diff
@@ -1031,40 +1021,64 @@ def forward(
+        b, np, sq, sk, hn = (
             query_layer.size(1),
             query_layer.size(2),
-            query_layer.size(0) - total_transient_tokens,
-            key_layer.size(0) - total_transient_tokens,
+            query_layer.size(0),
+            key_layer.size(0),
             query_layer.size(3),
         )
-
-        #attention_bias = torch.zeros(batch_size, np, sq, sk,
-        #                             dtype=query_layer.dtype,
-        #                             device=torch.cuda.current_device())
-
+                    
+        # ==================================================
+        # Update attention mask for inference. [b, np, sq, sk]
+        # ==================================================
+        if get_key_value:
+            with torch.no_grad():
+                if layer_past is not None:
+                    attention_mask = attention_mask[..., sq - 1, : sk].unsqueeze(2)
+                else:
+                    attention_mask = attention_mask[..., : sq, : sk]
+                    
+        # ==================================================
+        # Update attention bias. 
+        # relative_position_bias: [1, np, sq, sk]
+        # relative_position_bias w/ long-attention:  [1, np, nb, sq, sk]
+        # attention_bias: [b, np, sq, sk]
+        # ==================================================
+        attention_bias = None
         if self.position_embedding_type.lower() == 'sandwich':
-            relative_position_bias = sandwich_pos_bias(
+            attention_bias = sandwich_pos_bias(
                 sq, sk, self.hidden_size_per_attention_head, np, torch.cuda.current_device()
-            )[None]
-
+            )
+        
         if relative_position_bias is not None:
-            relative_position_bias = relative_position_bias[
-                :,
-                self.num_attention_heads_partition_offset : self.num_attention_heads_partition_offset
-                + self.num_attention_heads_per_partition,
-            ]
-            if relative_position_bias.shape[0] == 1:
-                relative_position_bias = relative_position_bias.expand(batch_size, *([-1]*(relative_position_bias.dim() - 1)))
-
-        # apply relative positional encoding (rotary embedding)
+            if isinstance(relative_position_bias, torch.Tensor):
+                attention_bias = relative_position_bias[
+                    :,
+                    self.num_attention_heads_partition_offset : self.num_attention_heads_partition_offset
+                    + self.num_attention_heads_per_partition,
+                ]
+                if attention_bias.shape[0] == 1:
+                    attention_bias = attention_bias.expand(b, *([-1]*(attention_bias.dim() - 1)))
+            else:
+                # relative_position_bias is T5RelativePositionEmbedding
+                attention_bias = self.get_local_pos_bias(
+                    relative_position_bias, nb, self.local_context
```

Comment:
nb isn't defined yet so this fails. though we are currently never using dynamic pos bias so maybe could remove this whole section for now